tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -4959,10 +4959,39 @@ static LogicalResult EliminateBroadcastInDimTranspose(
   return success();
 }
 
-void TransposeOp::getCanonicalizationPatterns(OwningRewritePatternList& results,
-                                              MLIRContext* /*context*/) {
+// Simplify Transpose: replace Transpose with Reshape if they are equivalent
+struct TransposeSimplifier : public OpRewritePattern<TransposeOp> {
+  using OpRewritePattern<TransposeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(TransposeOp op,
+                                PatternRewriter &rewriter) const override {
+    auto operand_type = op.operand().getType().dyn_cast<RankedTensorType>();
+    auto result_type = op.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!operand_type || !result_type) {
+      return failure();
+    }
+    if (!operand_type.hasStaticShape() || !result_type.hasStaticShape()) {
+      return failure();
+    }
+    auto permutation = op.permutation().getValues<int64_t>();
+    llvm::SmallVector<int64_t> sorted_permutation;
+    for (int64_t i = 0, e = result_type.getRank(); i < e; i++) {
+      if (result_type.getDimSize(i) != 1) {
+        sorted_permutation.push_back(permutation[i]);
+      }
+    }
+    if (llvm::is_sorted(sorted_permutation)) {
+      rewriter.replaceOpWithNewOp<ReshapeOp>(op, op.getType(), op.operand());
+      return success();
+    }
+    return failure();
+  }
+};
+
+void TransposeOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+                                              MLIRContext *context) {
   results.insert(EliminateRedundantTranspse);
   results.insert(EliminateBroadcastInDimTranspose);
+  results.insert<TransposeSimplifier>(context);
```

Comment:
Nit: since the other are using the function form right above, why not do the same?

Reply:
Done